Nonlinear regression simulations for split criteria comparison

morgsmss7 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->


#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Fixes Issue [16370](https://github.com/scikit-learn/scikit-learn/issues/16370) in scikit-learn. 

#### What does this implement/fix? Explain your changes.
This PR adds simulations and plots that show how split criteria compare on several nonlinear regression simulations including sinusoidal, logarithmic, multiplicative, and independence. There is not much information on scikit-learn's documentation about how to go about choosing which to use (mse, mae, or friedman mse) for the `criterion` parameter. This example demonstrates how to go about finding differences and shows that it may not always matter which criterion is chosen. 

#### Any other comments?
This PR will include these files:
```
sklearn/datasets/tests/test_samples_generator.py 
sklearn/datasets/samples_generator.py
sklearn/datasets/__init__.py
examples/ensemble/plot_random_forest_regression_criteria_comparison.py
examples/datasets/plot_nonlinear_regression_datasets.py 
doc/modules/classes.rst 
```
# Split Criteria Comparison Experiment and Results
For each simulation type (Logarithmic, Sine, Square, Multiplicative, Independence)
And For each split criterion (mse, mae, friedman mse)
1. Generate 30 noisy training sets (10 dimensions, 50 samples) 
2. Generate 1 noisy test set (10 dimensions, 1000 samples)
3. Train and evaluate using mse for all 30 training sets with random forests (500 trees) varying number of samples ( `np.arange(5, 51, 3)` )

![nonlinearSimPlots](https://user-images.githubusercontent.com/39866994/75188256-cf2f3d80-5719-11ea-9dc3-7ea001183931.png)
![splitter_comparison_02_17](https://user-images.githubusercontent.com/39866994/75184101-d3575d00-5711-11ea-9599-398d4b7155d9.png)

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
move all work from ndd branch relevant to regression criteria comparison to a new fork of sklearn to make PR

morgsmss7 committed:
Update classes.rst

morgsmss7 committed:
Update index.rst

morgsmss7 committed:
Merge pull request #1 from morgsmss7/regression-criteria-comparison

Nonlinear regression simulations for split criteria comparison

adrinjalali commented:
Thanks for the PR @morgsmss7 . You have linter issues, therefore the docs are not generated here for us to check how they look.

From a glance, this looks like a good start, but there are a few concerns I have:

- I'm not sure if we need to have those `dataset.make_*` methods. They're simple enough that they could go to the example needing them. But I'm happy to include them if other reviewers think it's a good thing to have.
    - If we include them, they'll need tests
    - I'm not sure if we need the example plotting the datasets. If we include it, then it may be useful to include other dataset generators as well? not sure.
- Our whole documentation is run and generated in less than 10 minutes. I'm not sure how long your example takes to run, but the runtime is definitely a concern.
- It's best to have the examples tell a story. You can check [this](https://scikit-learn.org/dev/auto_examples/inspection/plot_partial_dependence.html#sphx-glr-auto-examples-inspection-plot-partial-dependence-py) as a good example.

Please don't hesitate to ask if there are any questions.

Commit:
fix linting errors in criteria comparion .py file

Commit:
fix remaining linting issues in PR

Commit:
Merge branch 'regression-criteria-comparison' of https://github.com/morgsmss7/scikit-learn into regression-criteria-comparison

morgsmss7 committed:
Merge pull request #2 from morgsmss7/regression-criteria-comparison

Regression criteria comparison

Commit:
change .samples_generator to ._samples_generator and remove big print statements from nonlinear sim comparison

Commit:
added space after # for flake8

Commit:
reduce runtime to 3 minutes

Commit:
trying to fix sphinx errors

Commit:
fix flake8 issues

Commit:
fix nonlinear regression simulations tests

morgsmss7 commented:
@adrinjalali Thank you so much for the comments. I believe I fixed all of the linter issues. Does the documentation look okay to you?

I believe I've addressed most of your concerns:
1. The dataset.make_* methods have tests in sklearn/datasets/tests/test_samples_generator.py. There was a bug such that only one of them was tested multiple times. If you think it would be worth it to keep the data generation in samples_generator.py, I can include further tests. Otherwise, I can just move the data generation directly into examples/ensemble/plot_random_forest_regression_criteria_comparison.py and remove the tests. 
2. I believe I reduced the runtime significantly. Please let me know if it is still too long and I can decrease it further.
3. After looking at the documentation generated, I see that the output doesn't really tell a story. I will work on fleshing that out a bit more tonight.

Commit:
add more cells and explanations to help example to tell a story

Commit:
increase font size to make plots more readable

Commit:
fix flake8 issues

cmarmo commented:
Hi @morgsmss7, thanks for your work so far, and sorry for the late answer. If you are still interested in working on this, do you mind synchronizing with upstream? The rendering of your examples is no longer available on CircleCI, having them there will really help the reviewing process. Thank you for your patience!

lorentzenchr commented:
@morgsmss7 Thank you for this PR. Trying to clearify the usage of the different splitting criteria is very valuable. As the splitting criterion for decision trees is effectively the loss function being minimized, the following is not resteicted to trees: MSE should be used if you seek to predict the conditional (on X) expectation (mean value), MAE to predict the conditional median. To show a difference, one better uses non-symmetric target distributions, i.e. do not use the Normal distribution for y as expectation an median coincide.

The use of friedman_mse is unclear to me, see https://github.com/scikit-learn/scikit-learn/pull/17386#issuecomment-639079056. I could imagine to hide it from users in future versions. Better to exclude it from examples?

lorentzenchr commented:
@morgsmss7 I'll close as it seems stalled for a long time. In case you still want to continue, you can reopen, of course.
